bpo-32096: Remove obj and mem from _PyRuntime

vstinner opened the pull request:
bpo-32096, bpo-30860:  Partially revert the commit
2ebc5ce42a8a9e047e790aefbf9a94811569b2b6:

* Move structures back from Include/internal/mem.h to
  Objects/obmalloc.c
* Remove _PyObject_Initialize() and _PyMem_Initialize()
* Remove Include/internal/pymalloc.h

PyMem_RawMalloc() and Py_DecodeLocale() can be called again before
_PyRuntimeState_Init().

<!-- issue-number: bpo-32096 -->
https://bugs.python.org/issue32096
<!-- /issue-number -->

vstinner committed:
bpo-32096: Remove obj and mem from _PyRuntime

bpo-32096, bpo-30860:  Partially revert the commit
2ebc5ce42a8a9e047e790aefbf9a94811569b2b6:

* Move structures back from Include/internal/mem.h to
  Objects/obmalloc.c
* Remove _PyObject_Initialize() and _PyMem_Initialize()
* Remove Include/internal/pymalloc.h

PyMem_RawMalloc() and Py_DecodeLocale() can be called again before
_PyRuntimeState_Init().

vstinner committed:
Fix pgenmain.c

ncoghlan reviewed (dismissed):
While I'm +1 for this reversion, it should have a NEWS entry, since it's reverting changes that were published in previous alpha releases.

It should also incorporate @ericsnowcurrently's new test case from https://github.com/python/cpython/pull/4495/files#diff-559231a69a7d7f06ab48269b3bfa1f45, to confirm that it actually is fixing the regression it's intended to fix.

vstinner committed:
Fix _PyMem_GetDefaultRawAllocator()

vstinner committed:
Add NEWS entry

vstinner committed:
Add test_capi.test_pre_initialization_api()

Make sure that it's possible to call Py_DecodeLocale(), and then call
Py_SetProgramName() with the decoded string, before Py_Initialize().

Co-Authored-By: Eric Snow <ericsnowcurrently@gmail.com>

ncoghlan commented on Misc/NEWS.d/next/Core and Builtins/2017-11-24-01-13-58.bpo-32096.CQTHXJ.rst:
```diff
@@ -0,0 +1,4 @@
+Revert memory allocator changes in the C API: move structures back from
+_PyRuntime to Objects/obmalloc.c. The memory allocators are now initialized
+again statically, and so PyMem_RawMalloc() and Py_DecodeLocale() can be
```
Slight grammar tweak: "now initialized again" -> "once again initialized"

vstinner replied:
Fixed

ncoghlan reviewed (approved):
LGTM, thanks!

vstinner committed:
NEWS: Fix grammar

vstinner committed:
PCbuild: Remove Include/internal/pymalloc.h
